File: renderers/web_core/src/v0_9/state/data-model.ts
/*
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {Subscription as BaseSubscription} from '../common/events.js';
import {A2uiDataError} from '../errors.js';
import {
  batchWrite,
  effect,
  getValue,
  peekValue,
  setValue,
  signal,
  Signal,
} from '../reactivity/signals.js';

/**
 * Represents a reactive connection to a specific path in the data model.
 */
export interface DataSubscription<T> extends BaseSubscription {
  /**
   * The current value at the subscribed path.
   */
  readonly value: T | undefined;
}

function isNumeric(value: string): boolean {
  return /^(0|[1-9]\d*)$/.test(value);
}

/**
 * Keys forbidden in path resolution to prevent prototype pollution
 * vulnerabilities. Accessing or mutating these keys via object paths can allow
 * attackers to modify Object.prototype or Function.prototype.
 */
const FORBIDDEN_KEYS = new Set(['__proto__', 'constructor', 'prototype']);

/**
 * A standalone, observable data store representing the client-side state.
 * It handles JSON Pointer path resolution and subscription management.
 */
export class DataModel {
  private data: Record<string, unknown> = {};
  private readonly signals: Map<string, Signal<any>> = new Map();
  private readonly subscriptions: Set<() => void> = new Set(); // To track direct subscriptions for dispose

  /**
   * Creates a new data model.
   *
   * @param initialData The initial data for the model. Defaults to an empty
   *     object.
   */
  constructor(initialData: Record<string, unknown> = {}) {
    this.data = initialData;
  }

  /**
   * Retrieves a Preact Signal for a specific data path.
   *
   * This provides a reactive way to access a value. If the value at the path
   * changes via `set()`, the signal will automatically be updated.
   *
   * @param path The JSON pointer path to create or retrieve a signal for.
   * @returns A Preact Signal representing the value at the specified path.
   * @throws {A2uiDataError} If path is null, undefined, or contains forbidden
   *     segments (`__proto__`, `constructor`, `prototype`).
   */
  getSignal<T>(path: string): Signal<T | undefined> {
    const normalizedPath = this.normalizePath(path);
    if (!this.signals.has(normalizedPath)) {
      this.signals.set(normalizedPath, signal(this.get(normalizedPath)));
    }
    return this.signals.get(normalizedPath) as Signal<T | undefined>;
  }

  /**
   * Updates the model at the specific path and notifies all relevant signals.
   * If path is '/' or empty, replaces the entire root.
   *
   * Note on `undefined` values:
   * - For objects: Setting a property to `undefined` removes the key from the
   * object.
   * - For arrays: Setting an index to `undefined` sets that index to
   * `undefined` but preserves the array length (sparse array).
   *
   * @param path The JSON pointer path to set value for.
   * @param value The value to set at the specified path.
   * @returns This `DataModel` instance for chaining.
   * @throws {A2uiDataError} If path is null, undefined, invalid for
   *     arrays/primitives, or contains forbidden segments (`__proto__`,
   *     `constructor`, `prototype`).
   */
  set(path: string, value: any): this {
    if (path === null || path === undefined) {
      throw new A2uiDataError('Path cannot be null or undefined.');
    }

    if (path === '/' || path === '') {
      this.data = value;
      this.notifyAllSignals();
      return this;
    }

    const segments = this.parsePath(path);
    const lastSegment = segments.pop()!;

    // Only an absent root is replaced with a container. A primitive root is a
    // value the caller put there, and writing a path through it is the same
    // error as writing through a primitive at any other depth.
    if (this.data === undefined || this.data === null) {
      this.data = {};
    } else if (typeof this.data !== 'object') {
      throw new A2uiDataError(
        `Cannot set path '${path}': the data model root is a primitive value.`,
        path,
      );
    }
    let current: any = this.data;
    for (let i = 0; i < segments.length; i++) {
      const segment = segments[i];

      if (Array.isArray(current)) {
        if (!isNumeric(segment)) {
          throw new A2uiDataError(
            `Cannot use non-numeric segment '${segment}' on an array in path '${path}'.`,
            path,
          );
        }

        // If we encounter a primitive where a container is expected, we cannot
        // proceed. We allow undefined/null to be overwritten by a new
        // container.
        const index = parseInt(segment, 10);
        const val = current[index];
        if (val !== undefined && val !== null && typeof val !== 'object') {
          throw new A2uiDataError(
            `Cannot set path '${path}': segment '${segment}' is a primitive value.`,
            path,
          );
        }

        if (val === undefined || val === null) {
          const nextSegment = i < segments.length - 1 ? segments[i + 1] : lastSegment;
          current[index] = isNumeric(nextSegment) ? [] : {};
        }
      } else {
        // Ensure we only inspect own properties to prevent inherited
        // Object.prototype properties (e.g., toString) from being treated as
        // existing state.
        const hasOwnProp = Object.prototype.hasOwnProperty.call(current, segment);
        if (hasOwnProp) {
          const propVal = current[segment];
          // If we encounter a primitive where a container is expected, we
          // cannot proceed. We allow undefined/null to be overwritten by a new
          // container.
          if (propVal !== undefined && propVal !== null && typeof propVal !== 'object') {
            throw new A2uiDataError(
              `Cannot set path '${path}': segment '${segment}' is a primitive value.`,
              path,
            );
          }
        }

        if (!hasOwnProp || current[segment] === undefined || current[segment] === null) {
          const nextSegment = i < segments.length - 1 ? segments[i + 1] : lastSegment;
          current[segment] = isNumeric(nextSegment) ? [] : {};
        }
      }

      current = current[segment];
    }

    if (Array.isArray(current) && !isNumeric(lastSegment)) {
      throw new A2uiDataError(
        `Cannot use non-numeric segment '${lastSegment}' on an array in path '${path}'.`,
        path,
      );
    }

    if (value === undefined) {
      if (Array.isArray(current)) {
        current[parseInt(lastSegment, 10)] = undefined;
      } else {
        delete current[lastSegment];
      }
    } else {
      current[lastSegment] = value;
    }

    this.notifySignals(path);
    return this;
  }

  /**
   * Retrieves data at a specific JSON pointer path.
   *
   * @param path The JSON pointer path to read from.
   * @returns The value at the specified path, or undefined if not found.
   * @throws {A2uiDataError} If path is null, undefined, or contains forbidden
   *     segments (`__proto__`, `constructor`, `prototype`).
   */
  get(path: string): any {
    if (path === null || path === undefined) {
      throw new A2uiDataError('Path cannot be null or undefined.');
    }
    if (path === '/' || path === '') {
      return this.data;
    }

    const segments = this.parsePath(path);
    let current: any = this.data;
    for (const segment of segments) {
      if (current === undefined || current === null || typeof current !== 'object') {
        return undefined;
      }

      if (Array.isArray(current)) {
        if (!isNumeric(segment)) {
          return undefined;
        }
        const index = parseInt(segment, 10);
        if (index < 0 || index >= current.length) {
          return undefined;
        }
        current = current[index];
      } else {
        if (!Object.prototype.hasOwnProperty.call(current, segment)) {
          return undefined;
        }
        current = current[segment];
      }
    }
    return current;
  }

  /**
   * Subscribes to changes at the specified data path.
   *
   * This is a backwards-compatible layer using Preact Signals internally. It
   * allows listeners to be notified whenever the value at the specified path
   * (or any of its ancestors/descendants) changes.
   *
   * @param path The JSON pointer path to observe.
   * @param onChange A callback fired whenever the value changes.
   * @returns A `DataSubscription` containing the initial value and an
   *     `unsubscribe` method.
   * @throws {A2uiDataError} If path is null, undefined, or contains forbidden
   *     segments (`__proto__`, `constructor`, `prototype`).
   */
  subscribe<T>(path: string, onChange: (value: T | undefined) => void): DataSubscription<T> {
    const sig = this.getSignal<T>(path);
    let isSync = true;
    let currentValue = peekValue(sig);

    const dispose = effect(() => {
      const val = getValue(sig);
      currentValue = val;
      if (!isSync) {
        onChange(val);
      }
    });
    isSync = false;

    this.subscriptions.add(dispose);

    return {
      get value() {
        return currentValue;
      },
      unsubscribe: () => {
        dispose();
        this.subscriptions.delete(dispose);
      },
    };
  }

  /**
   * Clears all internal subscriptions.
   */
  dispose(): void {
    for (const dispose of this.subscriptions) {
      dispose();
    }
    this.subscriptions.clear();
    this.signals.clear();
  }

  private normalizePath(path: string): string {
    if (path.length > 1 && path.endsWith('/')) {
      return path.slice(0, -1);
    }
    return path || '/';
  }

  private parsePath(path: string): string[] {
    return path
      .split('/')
      .filter(p => p.length > 0)
      .map(rawSegment => {
        const segment = rawSegment.replace(/~([01])/g, (_, g) => (g === '1' ? '/' : '~'));
        if (FORBIDDEN_KEYS.has(segment)) {
          throw new A2uiDataError(`Forbidden path segment '${segment}' in path '${path}'.`, path);
        }
        return segment;
      });
  }

  private notifySignals(path: string): void {
    const normalizedPath = this.normalizePath(path);

    batchWrite(() => {
      this.updateSignal(normalizedPath);

      // Notify Ancestors
      let parentPath = normalizedPath;
      while (parentPath !== '/' && parentPath !== '') {
        parentPath = parentPath.substring(0, parentPath.lastIndexOf('/')) || '/';
        this.updateSignal(parentPath);
      }

      // Notify Descendants
      for (const subPath of this.signals.keys()) {
        if (this.isDescendant(subPath, normalizedPath)) {
          this.updateSignal(subPath);
        }
      }
    });
  }

  private updateSignal(path: string): void {
    const sig = this.signals.get(path);
    if (sig) {
      const val = this.get(path);
      if (Array.isArray(val)) {
        setValue(sig, [...val]);
      } else if (typeof val === 'object' && val !== null) {
        setValue(sig, {...val});
      } else {
        setValue(sig, val);
      }
    }
  }

  private notifyAllSignals(): void {
    batchWrite(() => {
      for (const path of this.signals.keys()) {
        this.updateSignal(path);
      }
    });
  }

  private isDescendant(childPath: string, parentPath: string): boolean {
    if (parentPath === '/' || parentPath === '') {
      return childPath !== '/';
    }
    return childPath.startsWith(parentPath + '/');
  }
}

File: dart/a2ui_core/lib/src/core/data_model.dart
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import '../primitives/data_path.dart';
import '../primitives/errors.dart';
import '../primitives/reactivity.dart';

/// The maximum list index that auto-vivification will expand to.
///
/// Prevents OOM from paths like `/data/999999999` which would otherwise
/// allocate a billion-element list.
const int maxAutoVivifyIndex = 10000;

/// A standalone, observable data store representing the client-side state.
/// It handles JSON Pointer path resolution and reactive signal management.
class DataModel {
  Object? _data;
  final Map<String, WeakReference<Signal<Object?>>> _signals = {};

  DataModel([Object? initialData]) : _data = initialData ?? <String, Object?>{};

  /// Synchronously gets data at a specific JSON pointer path.
  Object? get(String path) {
    final dataPath = DataPath.parse(path);
    if (dataPath.isEmpty) return _data;

    Object? currentNode = _data;
    for (final String segment in dataPath.segments) {
      if (currentNode == null) return null;
      if (currentNode is Map<String, Object?>) {
        currentNode = currentNode[segment];
      } else if (currentNode is List<Object?>) {
        final int? index = int.tryParse(segment);
        if (index == null || index < 0 || index >= currentNode.length) {
          return null;
        }
        currentNode = currentNode[index];
      } else {
        return null;
      }
    }
    return currentNode;
  }

  /// Updates data at a specific path and notifies subscribers.
  void set(String path, Object? value) {
    final dataPath = DataPath.parse(path);

    batch(() {
      if (dataPath.isEmpty) {
        _data = value;
      } else {
        _data ??= <String, Object?>{};
        Object? current = _data;
        for (var i = 0; i < dataPath.segments.length - 1; i++) {
          final String segment = dataPath.segments[i];
          final String nextSegment = dataPath.segments[i + 1];
          final isNextNumeric = int.tryParse(nextSegment) != null;

          if (current is Map<String, Object?>) {
            if (!current.containsKey(segment) || current[segment] == null) {
              current[segment] = isNextNumeric
                  ? <Object?>[]
                  : <String, Object?>{};
            }
            current = current[segment];
          } else if (current is List<Object?>) {
            final int? index = int.tryParse(segment);
            if (index == null) {
              throw A2uiDataError(
                "Cannot use non-numeric segment '$segment' on a list.",
                path: path,
              );
            }
            if (index < 0 || index > maxAutoVivifyIndex) {
              throw A2uiDataError(
                'List index out of bounds: $index (max $maxAutoVivifyIndex)',
                path: path,
              );
            }
            while (current.length <= index) {
              current.add(null);
            }
            if (current[index] == null) {
              current[index] = isNextNumeric
                  ? <Object?>[]
                  : <String, Object?>{};
            }
            current = current[index];
          } else {
            throw A2uiDataError(
              "Cannot set path '$path': intermediate segment '$segment' is a "
              'primitive.',
              path: path,
            );
          }
        }

        final String lastSegment = dataPath.segments.last;
        if (current is Map<String, Object?>) {
          if (value == null) {
            current.remove(lastSegment);
          } else {
            current[lastSegment] = value;
          }
        } else if (current is List<Object?>) {
          final int? index = int.tryParse(lastSegment);
          if (index == null) {
            throw A2uiDataError(
              "Cannot use non-numeric segment '$lastSegment' on a list.",
              path: path,
            );
          }
          if (index < 0 || index > maxAutoVivifyIndex) {
            throw A2uiDataError(
              'List index out of bounds: $index (max $maxAutoVivifyIndex)',
              path: path,
            );
          }
          while (current.length <= index) {
            current.add(null);
          }
          current[index] = value;
        } else {
          // The parent resolved to a primitive, so there is nothing to
          // write into. Dropping the write would hide a malformed path.
          throw A2uiDataError(
            "Cannot set path '$path': '$lastSegment' is a property of a "
            'primitive value.',
            path: path,
          );
        }
      }

      _notifyPathAndRelated(dataPath);
    });
  }

  /// Returns a [ReadonlySignal] for a specific path.
  /// Internally cached using a [WeakReference] to prevent leaks.
  ReadonlySignal<T?> watch<T>(String path) {
    final normalizedPath = DataPath.parse(path).toString();
    final WeakReference<Signal<Object?>>? ref = _signals[normalizedPath];
    if (ref != null) {
      final Signal<Object?>? sig = ref.target;
      if (sig != null) {
        return sig as ReadonlySignal<T?>;
      }
    }

    final Signal<T?> sig = signal<T?>(get(normalizedPath) as T?);
    _signals[normalizedPath] = WeakReference(sig as Signal<Object?>);
    _pruneSignals();
    return sig;
  }

  void _notifyPathAndRelated(DataPath dataPath) {
    final changedPath = dataPath.toString();
    final String changedDescendantPrefix = _descendantPrefix(changedPath);
    for (final String entryPath in _signals.keys.toList()) {
      if (changedPath == entryPath ||
          entryPath.startsWith(changedDescendantPrefix) ||
          changedPath.startsWith(_descendantPrefix(entryPath))) {
        _getAndNotify(entryPath);
      }
    }
  }

  static String _descendantPrefix(String path) => path == '/' ? '/' : '$path/';

  void _getAndNotify(String path) {
    final WeakReference<Signal<Object?>>? ref = _signals[path];
    if (ref == null) return;

    final Signal<Object?>? sig = ref.target;
    if (sig == null) {
      _signals.remove(path);
      return;
    }

    final Object? newValue = get(path);
    // A container mutated in place keeps its identity, so the live object
    // would compare equal and suppress the notification. Hand over a copy,
    // and let the signal's equality check suppress genuinely unchanged
    // values; notifying unconditionally would wake unaffected observers.
    // Match on the bare `Map` and `List` types: a caller may hand over a
    // `Map<dynamic, dynamic>`, which a bare `{}` literal and YAML both
    // produce, and a pattern naming the type arguments would miss it and
    // fall through to the no-copy branch -- losing the notification.
    sig.set(switch (newValue) {
      final Map<String, Object?> map => Map<String, Object?>.of(map),
      final Map<Object?, Object?> map => Map<Object?, Object?>.of(map),
      final List<Object?> list => List<Object?>.of(list),
      _ => newValue,
    });
  }

  void _pruneSignals() {
    _signals.removeWhere((key, ref) => ref.target == null);
  }

  void dispose() {
    _signals.clear();
  }
}
